### ISSUE TITLE
web: eliminate render-blocking resources on page load

### ISSUE THREAD

**Issue by aoldershaw** [concourse/concourse#5131]
Based on a [Lighthouse report](https://googlechrome.github.io/lighthouse/viewer/?gist=f2cbe304e59dfd99b3ba25a456e8e466) run on the dashboard of our CI deployment (https://ci.concourse-ci.org/), we can see that the performance isn't great - 5.5 seconds to first contentful paint.

Lighthouse suggests that we can achieve a good deal of savings by "eliminating render-blocking resources". Anything that is not required for first render should ideally be deferred (if possible/not too difficult).

<img width="890" alt="Screen Shot 2020-02-07 at 9 58 08 AM" src="https://user-images.githubusercontent.com/26583442/74039796-79743a80-4990-11ea-978e-683c74629c69.png">

## Which resources are necessary for initial render

* `main.css` - there's a case to be made for inlining critical styles and asynchronously loading the rest of the css, but given that our css is pretty small, it might not be worth the effort
* ~`index.js`~ - needed for rendering pipelines using d3. Could be dynamically imported if feasible
* ~`d3.v355.min.js`~ - needed for rendering pipelines into svgs. Could be dynamically imported if feasible
* ~`graph.js`~ - used by `index.js` to lay pipelines out. Could be dynamically imported or moved into `index.js` to have 1 fewer HTTP request
* ~`jquery-2.1.1.min.js`~ - I suspect we don't need to use jquery at all. We'll need to change some of the code in `index.js` and `elm-setup.js` to accomodate for this, though.
* `elm.min.js` - The compiled elm code is necessary for first render
* `elm-setup.js` - Sets up all the ports required for elm. Necessary for first render
*  `clipboard.min.js` - Probably not necessary, but Lighthouse doesn't seem to think the impact is large
* ~`mdi-svg.min.js`~ - needed for rendering icons on inputs/outputs in the pipeline graph. Could be dynamically imported. Additional note: somewhat counterintuitively, wrapping the massive `icon_paths` JSON object in a `JSON.parse(...)` (rather than treating it as a JS object literal) can improve performance: see https://www.youtube.com/watch?v=ff4fgQxPaO0

Note: by "dynamically importing", I had this feature in mind: https://v8.dev/features/dynamic-import#dynamic. It has decent browser support: https://caniuse.com/#feat=es6-module-dynamic-import. I'm definitely open to other suggestions for ways to achieve this.

**Comment by aoldershaw** [comment]
I'm running a local Concourse with 0 jobs/pipelines for my testing.

The initial Lighthouse results: 

<img width="975" alt="Screen Shot 2020-02-07 at 10 55 15 AM" src="https://user-images.githubusercontent.com/26583442/74044005-5d749700-4998-11ea-8d36-f9dcbfb7a34d.png">

<img width="924" alt="Screen Shot 2020-02-07 at 10 55 50 AM" src="https://user-images.githubusercontent.com/26583442/74044057-74b38480-4998-11ea-8b04-c4f10f49389f.png">

For the full results: https://googlechrome.github.io/lighthouse/viewer/?gist=d6d0c36bd7e11829963bc7fb79673903

**Comment by aoldershaw** [comment]
By dynamically importing only `mdi-svg.min.js`, we already see a substantial page load improvement:

<img width="792" alt="Screen Shot 2020-02-07 at 11 41 05 AM" src="https://user-images.githubusercontent.com/26583442/74047800-eb538080-499e-11ea-9b99-5da59d9b8a22.png">

For the full results: https://googlechrome.github.io/lighthouse/viewer/?gist=06ca8af419882ba0533ea37503730656

**Comment by aoldershaw** [comment]
Another attempt makes use of the `defer` attribute on script tags. `defer`'ing all "non-essential" scripts (`index.js`, `d3.v355.min.js`, `graph.js`, `jquery-2.1.1.min.js` - `mdi-svg.min.js` is still being dynamically imported) reduces the "First Contentful Paint" metric, but increases the "Time to Interactive" metric.

<img width="753" alt="Screen Shot 2020-02-07 at 12 33 22 PM" src="https://user-images.githubusercontent.com/26583442/74051713-7b48f880-49a6-11ea-97f7-b901fcfee92b.png">

This is because deferring downloads the scripts while parsing the rest of the HTML, but will then evaluate each of them after parsing the HTML. Only after evaluating all scripts on the main thread will the page become interactive. I'm not 100% sure why the TTI is larger here than in the benchmark (without any optimizations).

For the full results: https://googlechrome.github.io/lighthouse/viewer/?gist=fce9e37a57a2727bb33a25f0ce3b5438

**Comment by aoldershaw** [comment]
Initially I thought that, in order to dynamically import javascript, the file had to be an ES6 module (which would make it difficult to import something like d3), where our minified script simply sets globals). This isn't the case, however - if we just ignore the value resolved by the promise `import(...)`, upon the successful completion of the promise, we will have the required globals set. This means that fewer modifications need to be made.

**Comment by matthewpereira** [comment]
On a semi-related topic, does our elm build give us any opportunity to do tree-shaking on mdi-svg.js? 🤔

**Comment by aoldershaw** [comment]
> On a semi-related topic, does our elm build give us any opportunity to do tree-shaking on mdi-svg.js? 🤔

@matthewpereira I was thinking about that as well, but I don't think we can easily make any sort of optimization there, given that the icons to use are user defined (in the pipeline config): https://concourse-ci.org/resources.html#resource-icon

**Comment by pivotal-jamie-klassen** [comment]
@matthewpereira `mdi-svg.min.js` itself doesn't get built clean... though perhaps it should.
It's generated by the easily-forgotten `yarn update-mdi-svg`, which was introduced in #3581. The whole javascript file that gets generated is one big function `svg`, so in the strict sense of tree shaking there is no "dead code" -- as @aoldershaw was suggesting.

That said, it is quite bloated in that it contains the source of all the SVGs (the vast majority of which will not need to be rendered - there's stuff in there like polyhedral dice, weather and zodiac signs) under `node_modules/@mdi/svg/svg` -- we could shrink it drastically (like maybe remove it altogether) if we had the `publichandler` serving those SVGs as assets... they could perhaps be copied into the `public` directory as part of our RC builds.

@aoldershaw maybe worth thinking about, but I wouldn't sink too much time, in case it's rearranging deck chairs on the Titanic: one day https://project.concourse-ci.org/projects/MDc6UHJvamVjdDI5OTk4MjA= will hit the roadmap and we'll have a solid chance to revisit the relevant code paths.

**Comment by aoldershaw** [comment]
@pivotal-jamie-klassen I think getting rid of `mdi-svg.min.js` and requesting individual SVGs is the right approach, given that probably only a few icons will ever be used. For this issue, I'm just going to try to streamline the critical path as much as possible - and lazy-loading `mdi-svg.min.js` seems to work pretty well.

**Comment by aoldershaw** [comment]
With lazy-loading d3 and all of the graph rendering logic, we get further performance improvements:

<img width="743" alt="Screen Shot 2020-02-07 at 2 57 30 PM" src="https://user-images.githubusercontent.com/26583442/74061630-6f673180-49ba-11ea-8226-401b3eac17c4.png">

Full results: https://googlechrome.github.io/lighthouse/viewer/?gist=ff025076e076a599db183290d187253e

My next goal will be to remove all uses of jQuery in favour of native functions.

**Comment by pivotal-jamie-klassen** [comment]
> My next goal will be to remove all uses of jQuery in favour of native functions.

😎

**Comment by aoldershaw** [comment]
After replacing jQuery with native functions (mostly following https://github.com/nefe/You-Dont-Need-jQuery):

<img width="737" alt="Screen Shot 2020-02-07 at 4 11 28 PM" src="https://user-images.githubusercontent.com/26583442/74066186-a04c6400-49c4-11ea-8a71-4d5a0c5d4238.png">

In the green!

Full results: https://googlechrome.github.io/lighthouse/viewer/?gist=f9227f86e80bcc48aa34cde157dfc7e6

Not 100% ready to commit these changes since I'm not sure if I've introduced any regressions - that Javascript is pretty hard to follow.

**Comment by jomsie** [comment]
just jumping in here to say this is SICK!

**Comment by matthewpereira** [comment]
@aoldershaw This is awesome progress. Great work!

**Comment by kcmannem** [comment]
glad to see someone carrying on lightning mcQueens legacy
